sycl/plugins/level_zero/usm_allocator.cpp

```diff
@@ -0,0 +1,495 @@
+//===---------- usm_allocator.cpp - Allocator for USM memory --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <iostream>
+#include <list>
+#include <memory>
+#include <bitset>
+#include <algorithm>
+#include <cassert>
+#include <vector>
+#include <unordered_map>
+#include <utility>
+#include <mutex>
+#include <shared_mutex>
+
+#include "usm_allocator.hpp"
+
+namespace settings {
+  // Size of the slab which is going to be requested from the system.
+  static const size_t SlabSize = 64 * 1024; // 64K
+  // The largest size which is allocated via the allocator.
+  // Allocations with size > CutOff bypass the USM allocator and 
+  // go directly to the runtime.
+  static const size_t CutOff = SlabSize / 2;
+
+  // Generates a list of bucket sizes used by the allocator.
+  static std::vector<size_t> generateBucketSizes() {
+    std::vector<size_t> Sizes = {
+      32, 48, 64, 96, 128, 192
+    };
+
+    for (size_t size = 256; size < CutOff; size *= 2) {
+      Sizes.push_back(size);
+      Sizes.push_back(size + size / 2);
+    }
+
+    Sizes.push_back(CutOff);
+
+    return Sizes;
+  }
+
+  static const std::vector<size_t> BucketSizes = generateBucketSizes();
```

Comment:
Done, unfortunately I had to hardcode all the sizes instead of writing the loop due to some limitations of std::array(e.g. lack of constexpr operator[] for non-const std::array in C++14 , etc).